### PR TITLE
Use Java lookup rule for root of selection

### LINKED ISSUE #11788: Scala compiler conflicts Java field names with class / package canonical names
### Test example

**Foo.java**
```java
public class Foo {
    private String java;

    public java.lang.Integer test() {
        return null;
    }
}
```

**Bar.scala**
```scala
class Bar {
  def main(): Unit = {
    val bug = new Foo()
    bug.test()
  }
}
```

*compilation result:*
```
[error] /Users/fdaca/git/scala-bug-test/src/main/java/Foo.java:4:17: value lang is not a member of String
[error]     public java.lang.Integer test() {
[error]                 ^
```

### Bug description
Given a mixed Java / Scala project with a Java class having both:
- a method declaration with a canonical name of return type (e.g. `public java.lang.Integer test()`)
- a field with name equal to the first package of declared method return type (e.g. `private String java`)

Scala compiler would conflict the field identifier with package name and yield an error trying to lookup a member as if the first package was the previously declared field:  `value lang is not a member of String`.

The same problem occurs with different identifiers: `private Foo org;` + `public org.apache.commons.lang.CharSet getSet()` would try to lookup member `apache` in type `Foo`.

### Affected versions
- 2.13.1
- 2.12.8
- 2.11.8

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#11788

A follow-up to https://github.com/scala/scala/pull/8714

### CODE REVIEW

**Comment by lrytz** [comment]
jenkins is stuck, i'll take a look. but the failures on travis seem legit.

**Comment by som-snytt** [comment]
This commit is wrong, but I have a correction incoming.

File: test/files/pos/t11788/Bar.scala [hunk 1]
```diff
@@ -0,0 +1,3 @@
+object Bar extends App {
```
Comment (euartex) [inline test/files/pos/t11788/Bar.scala:1]: qqwq

File: test/files/neg/t11788b/java.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+
+public class java {
+	public static class lang {
```
Comment (euartex) [inline test/files/neg/t11788b/java.java:3]: qwwwqwqw

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 3]
```diff
@@ -5434,7 +5453,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             if (!isPastTyper && isUniversalMember(result.symbol))
               context.warning(tree.pos, s"dubious usage of ${result.symbol} with unit value", WarningCategory.LintUniversalMethods)
 
-          val sym = tree.symbol orElse member(qualTp, name) orElse inCompanionForJavaStatic(qualTp.typeSymbol, name)
+          val sym = tree.symbol
+            .orElse(member(qualTp, name))
```
Comment (euartex) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:5457]: sfg sgsfgsdg sfgdsdg

**Comment by som-snytt** [comment]
The Java spec says that for type `p.T`, check for class `p` with member `T`, then failing that, take `p` as a package.

This commit does that before failing a selection.

File: src/compiler/scala/tools/nsc/javac/JavaParsers.scala [hunk 4]
```diff
@@ -297,7 +297,7 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
         if (in.token == FINAL) in.nextToken()
         if (in.token == IDENTIFIER) {
           var t = typeArgs(atPos(in.currentPos)(Ident(ident())))
-          // typeSelect generates Select nodes is the lhs is an Ident or Select,
+          // typeSelect generates Select nodes if the lhs is an Ident or Select,
           // SelectFromTypeTree otherwise. See #3567.
```
Comment (euartex) [inline src/compiler/scala/tools/nsc/javac/JavaParsers.scala:301]: s sgsfgsdgdfsgsdfgfsg sfg sfs gfs sf

**Comment by som-snytt** [comment]
Blocking the user doesn't auto delete the graffiti?

**Comment by som-snytt** [comment]
The artifactory is 503.

**Comment by som-snytt** [comment]
Pushed a rebase and minor refactor of `repairJavaSelection`, and was richly rewarded.
```
java.io.IOException: No space left on device
```

**Comment by som-snytt** [comment]
I don't think it took, so I tried reverting a whitespace change I felt guilty about although it did actually look better. I was responsible for the previous version of the line about ensuring package is rooted.

**Comment by som-snytt** [comment]
actually the restored indentation is pretty cool

image

**Review by lrytz — APPROVED** [review]
very subtle!

**Pull request merged**